File: alsen_rx.py
import numpy   as np
from const import*


class alsen_rx(object):
   '''Приемник сигналов АЛСЕН'''
   def __init__(self):
      '''Инициализация'''
      self.Fcar = 174.89 # несущая АЛС-ЕН
      self.bit_rate = fmod
      self.fs = fs
      self.A = 1
      self.k = 2 * np.cos(2 * np.pi * self.Fcar / self.fs)

      self.X0_0 = 0
      self.X1_0 = 0
      self.X2_0 = self.A * np.sin(2 * np.pi * self.Fcar / self.fs)

      self.X0_90 = 0
      self.X1_90 = 0
      self.X2_90 = self.A * np.sin(2 * np.pi * self.Fcar / self.fs)
      self.cycle_count = 0

      self._buff_size = int((1/self.bit_rate)/(1/self.fs)/D)
      self._data0 = [0]*self._buff_size
      self._data90 = [0]*self._buff_size

   def local_gen(self,t):
      '''Локальный генератор cos и sin'''
      self.cycle_count = self.cycle_count + 1

      self.X0_0 = self.k*self.X1_0-self.X2_0
      y_0 = self.X0_0
      self.X2_0 = self.X1_0
      self.X1_0 = self.X0_0

      if self.cycle_count > 13:
         self.X0_90 = self.k*self.X1_90-self.X2_90
         y_90 = self.X0_90
         self.X2_90=self.X1_90
         self.X1_90=self.X0_90
         self.cycle_count = 13
      else:
         y_90 = 0

      return y_0, y_90

   def mux1(self,gen,x):
      '''Входные перемножители'''
      y = x * gen
      return y

   def diff_decode(self,x0,x90):
      '''Дифференциальный декодер'''

      y1 = x0  * self.delay0(x0)
      y2 = x90 * self.delay0(x0)
      y3 = x0  * self.delay90(x90)
      y4 = x90 * self.delay90(x90)

      y7 = (y2-y3) + (y1+y4) # Окунев стр 171
      y8 = (y1+y4) - (y2-y3)

      return y7, y8

   def delay0(self,x0):
      '''Задержка сигнала на длительность посылки канала 0'''
      self._data0.insert(0,x0)
      self._data0.pop()
      y0 = self._data0[self._buff_size-1]
      return y0

   def delay90(self,x90):
      '''Задержка сигнала на длительность посылки канала 90'''
      self._data90.insert(0,x90)
      self._data90.pop()
      y90 = self._data90[self._buff_size-1]
      return y90

   def sgn(self, sample):
      """Функция sgn"""
      return 0 if( sample > 0 or sample == 0) else 1

File: C++/alsen_rx.cpp
#include "alsen_rx.h"
#include <iostream>

alsen_rx::alsen_rx(const uint   ADeskrFreq,
                   const double AAmplitude,
                   const double ABitRate,
                   const double AFCar,
                   const uint   ADecimFactor)
    : Fcar(AFCar),          // несущая АЛС-ЕН
      bit_rate(ABitRate),   // скорость передачи Гц (Bod)
      fs(ADeskrFreq),       // частота дискретизации
      Ampl(AAmplitude),
      k(2 * cos(2 * M_PI * Fcar / fs)),

      X0_0(0.0),
      X1_0(0.0),
      X2_0(Ampl * sin(2 * M_PI * Fcar / fs)),

      X0_90(0.0),
      X1_90(0.0),
      X2_90(Ampl * sin(2 * M_PI * Fcar / fs)),
      cycle_count(0),

      DecimFactor(ADecimFactor), // фактор децимации
      buff_size(static_cast<int>((1.0/bit_rate)/(1.0/fs)/DecimFactor)),
      data0(buff_size),
      data90(buff_size)
{
    //Q_UNUSED(AAmplitude)
}

int alsen_rx::sgn(const double sample) const
{
    return (sample < 0) ? 1 : 0;
}

void alsen_rx::local_gen()
{
    cycle_count++;

    X0_0 = k * X1_0 - X2_0;
    double y_0 = X0_0;
    X2_0 = X1_0;
    X1_0 = X0_0;

    double y_90 = 0.0;

    if (cycle_count > 13)
    {
       X0_90 = k * X1_90 - X2_90;
       y_90 = X0_90;
       X2_90 = X1_90;
       X1_90 = X0_90;
       cycle_count = 13;
    }

    FResult.first = y_0;
    FResult.second = y_90;
}

double alsen_rx::mux1(const double gen, const double x) const
{
    return gen * x;
}

void alsen_rx::diff_decode(const double x0, const double x90)
{
    double y1 = x0  * delay0(x0);
    double y2 = x90 * delay0(x0);
    double y3 = x0  * delay90(x90);
    double y4 = x90 * delay90(x90);

    FResult.first = (y2 - y3) + (y1 + y4); //# Окунев стр 171
    FResult.second = (y1 + y4) - (y2 - y3);
}

double alsen_rx::delay0(const double x0)
{
    data0.insert(data0.begin(),x0);
    data0.pop_back();
    return data0[data0.size()-1];
}

double alsen_rx::delay90(const double x90)
{
    data90.insert(data90.begin(),x90);
    data90.pop_back();
    return data90[data90.size()-1];
}
